src/index.ts

```diff
@@ -117,16 +121,36 @@ export class Pipeline implements Serializable {
       ? {
           explicitDependencies: true,
           cache: new Map(),
+          mutator: opts.mutator,
         }
       : {
           explicitDependencies: false,
+          mutator: opts.mutator,
         };
 
+    const steps = await this.toList(newOpts);
+    if (opts.mutator) {
+      for (const step of steps) {
+        if (step instanceof Step) {
+          const deps = {
+            dependencies: new Set(step.dependencies),
+            effectDependencies: new Set(step.effectDependencies),
+          };
+          await opts.mutator(step);
+          if (
```

Comment:
we can use `ok` from the node `assert` here:

```
import { ok } from 'assert';

ok(isEqual(deps, { ...}), 'mutator must not mutate dependencies or effects'); 
```

Reply:
done